ArduCopter/mode_autotune.cpp

```diff
@@ -1390,36 +1446,45 @@ void Copter::ModeAutoTune::updating_p_up(float &tune_p, float tune_p_max, float
+// updating_angle_p_down - decrease P until we don't reach the target before time out
+// P is decreased to ensure we are not overshooting the target
+void Copter::ModeAutoTune::updating_angle_p_down(float &tune_p, float tune_p_min, float tune_p_step_ratio, float angle_target, float meas_angle_max, float meas_rate_min, float meas_rate_max)
 {
-    if (measurement_max > target*(1+0.5f*g.autotune_aggressiveness)) {
+    if (meas_angle_max < angle_target*(1+0.5f*g.autotune_aggressiveness)) {
+        if (ignore_next == false) {
+            // if maximum measurement was lower than target so increment the success counter
+            counter++;
+        } else {
+            ignore_next = false;
+        }
+    }else{
         // ignore the next result unless it is the same as this one
         ignore_next = true;
-        // if maximum measurement was greater than target so increment the success counter
-        counter++;
-    } else if ((measurement_max < target) && (measurement_max > target*(1.0f-AUTOTUNE_D_UP_DOWN_MARGIN)) && (measurement_max-measurement_min > measurement_max*g.autotune_aggressiveness) && (tune_d > tune_d_min)) {
-        // if bounce back was larger than the threshold so decrement the success counter
+        // if maximum measurement was higher than target so decrement the success counter
         if (counter > 0 ) {
             counter--;
         }
-        // decrease D gain (which should decrease bounce back)
-        tune_d -= tune_d*tune_d_step_ratio;
-        // stop tuning if we hit minimum D
-        if (tune_d <= tune_d_min) {
-            tune_d = tune_d_min;
-            Log_Write_Event(DATA_AUTOTUNE_REACHED_LIMIT);
-        }
-        // decrease P gain to match D gain reduction
+        // decrease P gain (which should decrease the maximum)
         tune_p -= tune_p*tune_p_step_ratio;
-        // stop tuning if we hit minimum P
+        // stop tuning if we hit maximum P
         if (tune_p <= tune_p_min) {
             tune_p = tune_p_min;
+            counter = AUTOTUNE_SUCCESS_COUNT;
             Log_Write_Event(DATA_AUTOTUNE_REACHED_LIMIT);
         }
-        // cancel change in direction
-        positive_direction = !positive_direction;
+    }
+}
+
+// updating_angle_p_up - increase P to ensure the target is reached
+// P is increased until we achieve our target within a reasonable time
+void Copter::ModeAutoTune::updating_angle_p_up(float &tune_p, float tune_p_max, float tune_p_step_ratio, float angle_target, float meas_angle_max, float meas_rate_min, float meas_rate_max)
+{
+    if ((meas_angle_max > angle_target*(1+0.5f*g.autotune_aggressiveness)) ||
+            ((meas_angle_max > angle_target) && (meas_rate_min < -meas_rate_max*g.autotune_aggressiveness))) {
+        // ignore the next result unless it is the same as this one
+        ignore_next = 1;
```

Comment:
Shouldn't this be `ignore_next = true` rather then using 1? (I get that the compiler will fix it, but I'm guessing it will also complain about it, haven't checked the CI output)

Reply:
That has been in there for a long time. I will fix it. Thanks!!